diff --git a/arch/arm/boot/dts/sun8i-h3-orangepi-pc-plus.dts b/arch/arm/boot/dts/sun8i-h3-orangepi-pc-plus.dts
index 948df1868..3b9426ab4 100644
--- a/arch/arm/boot/dts/sun8i-h3-orangepi-pc-plus.dts
+++ b/arch/arm/boot/dts/sun8i-h3-orangepi-pc-plus.dts
@@ -49,7 +49,16 @@
 
 	aliases {
 		/* ethernet0 is the H3 emac, defined in sun8i-h3.dtsi */
-		ethernet1 = &rtl8189ftv;
+		ethernet0 = &emac;
+	};
+	reg_gmac_3v3: gmac-3v3 {
+		compatible = "regulator-fixed";
+		regulator-name = "gmac-3v3";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		startup-delay-us = <100000>;
+		enable-active-high;
+		gpio = <&pio 3 6 GPIO_ACTIVE_HIGH>;
 	};
 
 	wifi_pwrseq: wifi_pwrseq {
@@ -59,8 +68,21 @@
 };
 
 &emac {
-	/* LEDs changed to active high on the plus */
-	/delete-property/ allwinner,leds-active-low;
+	pinctrl-names = "default";
+	pinctrl-0 = <&emac_rgmii_pins>;
+	phy-supply = <&reg_gmac_3v3>;
+	phy-handle = <&ext_rgmii_phy>;
+	phy-mode = "rgmii";
+
+	allwinner,leds-active-low;
+	status = "okay";
+};
+
+&external_mdio {
+	ext_rgmii_phy: ethernet-phy@1 {
+		compatible = "ethernet-phy-ieee802.3-c22";
+		reg = <0>;
+	};
 };
 
 &mmc1 {

diff --git a/drivers/char/Kconfig b/drivers/char/Kconfig
index c28dca0c..a2cfab25 100644
--- a/drivers/char/Kconfig
+++ b/drivers/char/Kconfig
@@ -7,6 +7,10 @@ menu "Character devices"
 
 source "drivers/tty/Kconfig"
 
+config DR_FIX
+	bool "Devesh's Fix for the H5 USB Regulators"
+	default y
+
 config DEVMEM
 	bool "/dev/mem virtual device support"
 	default y
diff --git a/drivers/char/Makefile b/drivers/char/Makefile
index 7dc3abe6..adca10ad 100644
--- a/drivers/char/Makefile
+++ b/drivers/char/Makefile
@@ -15,6 +15,7 @@ obj-$(CONFIG_UV_MMTIMER)	+= uv_mmtimer.o
 obj-$(CONFIG_IBM_BSR)		+= bsr.o
 obj-$(CONFIG_SGI_MBCS)		+= mbcs.o
 obj-$(CONFIG_BFIN_OTP)		+= bfin-otp.o
+obj-$(CONFIG_DR_FIX)		+= dr-fix.o
 
 obj-$(CONFIG_PRINTER)		+= lp.o

diff --git a/drivers/char/dr-fix.c b/drivers/char/dr-fix.c
new file mode 100644
index 00000000..2187fe87
--- /dev/null
+++ b/drivers/char/dr-fix.c
@@ -0,0 +1,41 @@
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/gpio.h>                 // Required for the IO functions
+#include <linux/interrupt.h>            // Required for the IRQ code
+
+#warning ("Building Devesh's Fix for the Firmware.")
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Devesh Rai");
+MODULE_DESCRIPTION("Fix for USB Regulator");
+MODULE_VERSION("0.1");
+static unsigned int regulator1 = 128;    // For USB2-DRVVBUS (PE0/128)  
+static unsigned int regulator2 = 355;   // For USB1-DRVVBUS (PL3/355)
+//static unsigned int regulator3 = 354;   // For USB0-DRVVBUS (PL2/354)
+
+
+static int __init dr_fix_init(void){
+   printk("Devesh's Fix for the USB Regulator.");
+   gpio_request(regulator1, "sysfs");          // USB2-DRVVBUS is hardcoded to 128, request it
+   gpio_direction_output(regulator1, true);   // Set the gpio to be in output mode and on
+   gpio_set_value(regulator1, true);          // Not required as set by line above (here for reference)
+
+   gpio_request(regulator2, "sysfs");          // USB1-DRVVBUS is hardcoded to 355, request it
+   gpio_direction_output(regulator2, true);   // Set the gpio to be in output mode and on
+   gpio_set_value(regulator2, true);          // Not required as set by line above (here for reference)
+/*
+   gpio_request(regulator3, "sysfs");          // USB0-DRVVBUS is hardcoded to 354, request it
+   gpio_direction_output(regulator3, true);   // Set the gpio to be in output mode and on
+   gpio_set_value(regulator3, true);          // Not required as set by line above (here for reference)
+*/
+   return 1;
+   
+}
+
+static void __exit dr_fix_exit(void){
+   printk(KERN_INFO "Goodbye!\n");
+}
+
+module_init(dr_fix_init);
+module_exit(dr_fix_exit);
